File: samples/FsBulletML.Sample.Unity2D.CSharp/Assets/Scripts/EnemyBullet.cs
using UnityEngine;
using System.Collections;
using Microsoft.FSharp.Core;
using FsBulletML;

public class EnemyBullet : BaseBullet
{
    public EnemyBullet()
        : base()
    {
        var self = this as FsBulletML.Processable.IBulletmlObject;
        self.Init();
        this.Root = true;

        self.IsBullet = true;
        self.BulletRoot = true;
        self.BulletType = Processable.BulletType.Enemy;
    }

    protected new void Update()
    {
        base.Update();

        var self = this as FsBulletML.Processable.IBulletmlObject;
        if (!this.Root && self.BulletRoot && !self.Used)
        {
            InstanceManager.Destroy(gameObject);
        }

        if (this.transform.position.x < 0 || this.transform.position.x > 4.8)
        {
            self.Used = false;
            InstanceManager.Destroy(gameObject);
        }


        if (this.transform.position.y < -6.4 || this.transform.position.y > 0)
        {
            self.Used = false;
            InstanceManager.Destroy(gameObject);
        }
    }

    public override GameObject GetBulletPrefubInstance()
    {
        return InstanceManager.InstantiatePrefab(this.bulletObject, this.transform.position, this.transform.rotation);
    }

    public void SetTask(FSharpOption<Processable.BulletmlTask> bulletmlTask)
    {
        var self = this as FsBulletML.Processable.IBulletmlObject;
        self.Task = bulletmlTask;
    }

    void OnTriggerEnter2D(Collider2D collier)
    {
        if (collier.gameObject.tag == "Player")
        {
            if (!this.Root)
            {
                InstanceManager.Destroy(gameObject);
            }
        }
    }

}

File: samples/FsBulletML.Sample.Unity2D.FSharp/FsBulletML.Sample.Unity2D.FSharp/EnemyBullet.fs
namespace FsBulletML.Sample.Unity2D.FSharp

open System
open System.Collections.Generic
open System.Runtime.Serialization
open UnityEngine
open FsBulletML
open FsBulletML.Unity2D 
 
type EnemyBullet () =
  inherit BaseBullet ()

  member this.Awake () =
    base.Awake ()
    let self = this.GetDefaultBullet ()
    self.Init()
    self.Root <- true
    self.IsBullet <- true
    self.BulletRoot <- true
    self.BulletType <- BulletType.Enemy 

  override this.Update () = 
    base.Update()
    let self = this.GetDefaultBullet ()
    if (not self.Root && self.BulletRoot && not self.Used) then
      InstanceManager.Destroy(this.gameObject)

    if (this.transform.position.x < 0.f || this.transform.position.x > 4.8f) then
      self.Used <- false
      InstanceManager.Destroy(this.gameObject)

    if (this.transform.position.y < -6.4f || this.transform.position.y > 0.f) then
      self.Used <- false
      InstanceManager.Destroy(this.gameObject)

  override this.GetBulletPrefubInstance () =
    InstanceManager.InstantiatePrefab(this.bulletObject, this.transform.position, this.transform.rotation)

  member this.SetTask(bulletmlTask) = 
    let self = this.GetDefaultBullet ()
    self.Task <- bulletmlTask

  member this.OnTriggerEnter2D(collier:Collider2D) =
    if (collier.gameObject.tag = "Player") then
      if (not <| this.GetDefaultBullet().Root) then
        InstanceManager.Destroy(this.gameObject)